src/pykeen/datasets/analysis.py

```diff
@@ -0,0 +1,449 @@
+    >>> df.loc['training', ('head', 'embassy')].sort_values().tail()
+
+    # In which countries are to most embassies (considering only training triples)?
+    >>> df.loc['training', ('tail', 'embassy')].sort_values().tail()
+
+    :param dataset:
+        The dataset.
+
+    :return:
+        A dataframe with a multi-index (subset, entity_id) as index, and a multi-index (kind, relation) as columns,
+        where subset in {'training', 'validation', 'testing', 'total'}, and kind in {'head', 'tail'}. For each entity,
+        the corresponding row can be seen a pseudo-type, i.e. for which relations it may occur as head/tail.
+    """
+    num_relations = dataset.num_relations
+    num_entities = dataset.num_entities
+    data = numpy.zeros(shape=(4 * num_entities, 2 * num_relations), dtype=numpy.int64)
+    for i, (_subset_name, triples_factory) in enumerate(sorted(dataset.factory_dict.items())):
+        # head-relation co-occurrence
+        unique_hr, counts_hr = triples_factory.mapped_triples[:, :2].unique(dim=0, return_counts=True)
+        h, r = unique_hr.t().numpy()
+        data[i * num_entities:(i + 1) * num_entities, :num_relations][h, r] = counts_hr.numpy()
+
+        # tail-relation co-occurrence
+        unique_rt, counts_rt = triples_factory.mapped_triples[:, 1:].unique(dim=0, return_counts=True)
+        r, t = unique_rt.t().numpy()
+        data[i * num_entities:(i + 1) * num_entities, num_relations:][t, r] = counts_rt.numpy()
+
+    # full dataset
+    data[3 * num_entities:] = sum(data[i * num_entities:(i + 1) * num_entities] for i in range(3))
+    entity_id_to_label, relation_id_to_label = [
+        invert_mapping(mapping=mapping)
+        for mapping in (dataset.entity_to_id, dataset.relation_to_id)
+    ]
+    return pandas.DataFrame(
+        data=data,
+        index=pandas.MultiIndex.from_product([
+            sorted(dataset.factory_dict.keys()) + ['total'],
+            [entity_id_to_label[entity_id] for entity_id in range(num_entities)],
+        ]),
+        columns=pandas.MultiIndex.from_product([
+            ('head', 'tail'),
+            [relation_id_to_label[relation_id] for relation_id in range(num_relations)],
+        ]),
+    )
+
+
+def _get_skyline(xs: Collection[Tuple[int, float]]) -> Collection[Tuple[int, float]]:
+    # TODO: naive implementation, O(n2)
+    return {
+        (s, c)
+        for s, c in xs
+        if not any(
+            s2 > s and c2 > c
+            for s2, c2 in xs
+        )
+    }
+
+
+def skyline(
+    data_stream: Iterable[Tuple[int, str, int, float]],
+) -> Iterable[Tuple[int, str, int, float]]:
+    """
+    Keep only those entries which are in the support-confidence skyline.
+
+    A pair (s, c) dominates (s', c') if s > s' and c > c'. The skyline contains those entries which are not dominated
+    by any other entry.
+
+    :param data_stream:
+        The stream of data, comprising tuples (relation_id, pattern-type, support, confidence).
+
+    :return:
```

Comment:
@mberr you want to write instead 
```
yields:
```
and then describe a single object that gets yielded (since it's implied several will be yielded by the type annotation)

Reply:
with or without another `:` in front? :sweat_smile:

Reply:
my b. That was a typo. Use the `:yields:`